pkg/database/authorized_app_stats.go

```diff
@@ -0,0 +1,191 @@
+		// All entires are 30d
+		summary.CodesIssued30d += AuthorizedAppStats.CodesIssued
+
+		// Only one entry is 1d
+		if AuthorizedAppStats.Date == roundedTime {
+			summary.CodesIssued1d += AuthorizedAppStats.CodesIssued
+		}
+
+		// Find 7d entries
+		if AuthorizedAppStats.Date.After(roundedTime.AddDate(0, 0, -7)) {
+			summary.CodesIssued7d += AuthorizedAppStats.CodesIssued
+
+		}
+	}
+
+	// create 24h, 7d, 30d counts
+	return summary, nil
+}
+
+func (db *Database) GetAuthorizedAppStats(a *AuthorizedApp, r *Realm, t time.Time) (*AuthorizedAppStats, error) {
+	var AuthorizedAppStats AuthorizedAppStats
+	roundedTime := t.Truncate(24 * time.Hour)
+
+	if err := db.db.
+		Where("authorized_app_id = ?", a.ID).
+		Where("realm_id = ?", r.ID).
+		Where("date = ?", roundedTime).
+		First(&AuthorizedAppStats).Error; err != nil {
+		return nil, err
+	}
+	return &AuthorizedAppStats, nil
+}
+
+func (db *Database) UpdateAuthorizedAppStats() error {
+	var appStats AuthorizedAppStats
+	err := db.db.Order("date DESC").First(&appStats).Error
+
+	if err != nil {
+		return err
+	}
+
+	// Start from last day in authorized app table.
+	curDay := appStats.Date
+	t := time.Now().UTC()
+	now := time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, time.UTC)
+
+	for {
+		if curDay.After(now) {
+			return nil
+		}
+
+		err := db.updateAuthorizedAppStatsDay(curDay)
+		if err != nil {
+			return err
+		}
+		curDay = curDay.AddDate(0, 0, 1)
+	}
+
+}
+
+func (db *Database) updateAuthorizedAppStatsDay(t time.Time) error {
+	roundedTime := t.Truncate(24 * time.Hour)
+
+	// For each realm, and each user in the realm, gather and store stats
+	realms, err := db.GetRealms()
+	if err != nil {
+		return err
+	}
+
+	for _, realm := range realms {
+		apps, err := realm.GetAuthorizedApps(db, false)
+		if err != nil {
+			return err
+		}
+
+		for _, app := range apps {
+			codesIssued, err := db.countVerificationCodesByAuthorizedApp(realm.ID, app.ID, roundedTime)
+			if err != nil {
+				return err
+			}
+
+			var appStats AuthorizedAppStats
+			err = db.db.
+				Where("authorized_app_id = ?", app.ID).
+				Where("realm_id = ?", realm.ID).
+				Where("date = ?", roundedTime).
+				First(&appStats).Error
+
+			if err == gorm.ErrRecordNotFound {
+				// New record.
+				appStats = AuthorizedAppStats{}
+			} else if err != nil {
+				return err
+			}
+
+			appStats.Date = roundedTime
+			appStats.RealmID = realm.ID
+			appStats.AuthorizedAppID = app.ID
+			appStats.CodesIssued = codesIssued
+
+			if appStats.Model.ID == 0 {
+				if err := db.db.Create(&appStats).Error; err != nil {
+					return err
+				}
+			} else {
+				if err := db.db.Save(&appStats).Error; err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (db *Database) countVerificationCodesByAuthorizedApp(realmID uint, appID uint, t time.Time) (uint64, error) {
+	if appID <= 0 {
+		return 0, nil
+	}
+
+	var rows []*VerificationCode
+	if err := db.db.Model(&VerificationCode{}).
```

Comment:
Using `.Count` would push this into the SQL layer and would probably be faster (since we wouldn't have to allocate all the rows just to count them)

Reply:
Count would be faster but count requires locking, which was a previous concern raised, I can go back to that certainly.

Reply:
https://github.com/google/exposure-notifications-verification-server/issues/200 to track